### ISSUE TITLE
RuntimeError: CUDA error: the provided PTX was compiled with an unsupported toolchain.

### ISSUE THREAD

**Issue by Clevrai** [googlecolab/colabtools#2452]
We are attempting to train a neural net using Colab's GPU offerings. When getting paired with an A100 GPU, we experience the following issue:

```
RuntimeError: CUDA error: the provided PTX was compiled with an unsupported toolchain.
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
```

To verify that it wasn't something to do with the actual program we were running, we ran the same program on a GCP instance and with a Colab V100. In both cases, the program ran fine without issue/as expected. 

To further verify that the program wasn't the issue, we ran the following code snippet:

```
import torch
torch.tensor([0,1],device="cuda")
```

And had the same issue. 

Is anyone else experiencing this issue?

**Comment by Snnzhao** [comment]
I also meet the same problem.

**Comment by Snnzhao** [comment]
https://github.com/rusty1s/pytorch_scatter/issues/225
Here may be the solution.

**Comment by Snnzhao** [comment]
Well the problem is not solved through above solution.
Again I firt meet this problem on A100 yesterday. Before that, all thing runs well with the same code and environment configuratioin.

**Comment by morgangiraud** [comment]
Same here with JAX.

**Comment by JooblesTheGreat** [comment]
On PyTorch and can confirm that I get the same error on the A100s but not on the V100s. Just a few days ago my code ran fine on both Gpus, but now it won't work.

> RuntimeError: CUDA error: the provided PTX was compiled with an unsupported toolchain.

**Comment by Clevrai** [comment]
Good to know we weren't the only ones! If it's affecting JAX as well then it must be at the GPU level. We submitted a request to Colab through their feedback forum and hopefully, it gets there quicker.

**Comment by craigcitro** [comment]
Thanks for all the reports!

From a quick experiment, here's what I'm seeing ([sample notebook](https://colab.research.google.com/gist/craigcitro/24b25055a0583da98a8a835d660fe024/b-207675214-scratch.ipynb)):
* the preinstalled versions of TF and torch work fine for me on an A100
* the preinstalled version of JAX is not working

Internal tracking: b/207675214

Questions for users who commented above:
* @Clevrai did you install your own version of torch? (It's way too easy to accidentally get a mismatch between installed CUDA versions, esp. if installing from PyPI.)
* @Snnzhao what library were you using?
* @morgangiraud were you using the preinstalled JAX?

**Comment by Clevrai** [comment]
@craigcitro Let us try one more time and see if we get the same issue. As far as we're aware, we were using the default torch that is installed with Colab.

**Comment by craigcitro** [comment]
@Clevrai Thanks for the fast reply! If you can cook up a minimal repro with the default-installed torch, I'm happy to dig a little more on our side.

**Comment by Clevrai** [comment]
@craigcitro No worries! For the time being, we're using the library "SimpleT5" as the project (`pip install simplet5`). I don't believe this is the issue given it doesn't have any trouble with the other runtimes (V100, GCP instance). The repo has the following requirements:

`
  
pandas
sentencepiece
torch>=1.7.0,!=1.8.0
transformers==4.10.0
pytorch-lightning==1.4.5
`

Again, we just used this the other day without issues. Does this give you issues?

**Comment by Clevrai** [comment]
@craigcitro Still get the same issue even if I clone the repo (i.e: without uninstalling the version of torch).

**Comment by craigcitro** [comment]
I did a `!pip install simplet5`, restarted my runtime, and was still able to successfully import torch and run basic snippets, eg:

```
device = torch.cuda.current_device()
x = torch.randn(70000, 16).to(device)
y = torch.randn(16, 70000).to(device)
z = torch.matmul(x, y)
z.shape, z.device
```

returns `(torch.Size([70000, 70000]), device(type='cuda', index=0))`.

Can you share a minimal repro notebook for the error you're encountering?

**Comment by Clevrai** [comment]
Yes, here's what I have:

`
! pip install simplet5

# Works fine
import pandas as pd
from sklearn.model_selection import train_test_split

df = pd.read_csv('p3.csv')

df = df.rename(columns={"target_text":"target_text", "input_text":"source_text"})
df = df[['source_text', 'target_text']]

# Works
from simplet5 import SimpleT5

model = SimpleT5()
model.from_pretrained(model_type="t5", model_name="t5-base")

# Fails with the error
model.train(train_df=train_df,
            eval_df=test_df, 
            source_max_token_len=512, 
            target_max_token_len=256, 
            batch_size=21,
            max_epochs=5, 
            early_stopping_patience_epochs=1,
            use_gpu=True
            )

`

**Comment by morgangiraud** [comment]
@craigcitro I check the dependencies from a `setup.py` file with the following requirement: `jax>=0.2.25`

The output is:
```
Requirement already satisfied: jax>=0.2.25 in /usr/local/lib/python3.7/dist-packages (from ...) (0.2.25)
Requirement already satisfied: jaxlib>=0.1.69 in /usr/local/lib/python3.7/dist-packages (from ...) (0.1.74+cuda11.cudnn805)
```

So I guess I'm using the preinstalled JAX.

**Comment by craigcitro** [comment]
@Clevrai do you have a repro with publicly available data? (I'd like something I could run on my side.)

@morgangiraud can you try running `!pip install https://storage.googleapis.com/jax-releases/cuda111/jaxlib-0.1.71+cuda111-cp37-none-manylinux2010_x86_64.whl` and restarting your runtime? (That will install our previous version of jaxlib.) I'm talking with the jax folks to confirm this doesn't set you up for further issues, but may get you unblocked.

**Comment by Clevrai** [comment]
@craigcitro  Yes, if you run the following snippet, I get the same error:

```
! pip install simplet5


import pandas as pd
from sklearn.model_selection import train_test_split

path = "https://raw.githubusercontent.com/Shivanandroy/T5-Finetuning-PyTorch/main/data/news_summary.csv"
df = pd.read_csv(path)


# simpleT5 expects dataframe to have 2 columns: "source_text" and "target_text"
df = df.rename(columns={"headlines":"target_text", "text":"source_text"})
df = df[['source_text', 'target_text']]

# T5 model expects a task related prefix: since it is a summarization task, we will add a prefix "summarize: "
df['source_text'] = "summarize: " + df['source_text']

train_df, test_df = train_test_split(df, test_size=0.2)




from simplet5 import SimpleT5

model = SimpleT5()
model.from_pretrained(model_type="t5", model_name="t5-base")
model.train(train_df=train_df,
            eval_df=test_df, 
            source_max_token_len=128, 
            target_max_token_len=50, 
            batch_size=8, max_epochs=3, use_gpu=True)

```

I get the same error.

**Comment by Clevrai** [comment]
@morgangiraud If you run my above snippet, do you get the same issue?

**Comment by Clevrai** [comment]
@craigcitro Slight odd behavior, when I don't restart the runtime with an A100 and run the snippet:

```
device = torch.cuda.current_device()
x = torch.randn(70000, 16).to(device)
y = torch.randn(16, 70000).to(device)
z = torch.matmul(x, y)
z.shape, z.device
```
I get:

```
(torch.Size([70000, 70000]), device(type='cuda', index=0)
```
But if I don't restart the runtime, I get:

```
RuntimeError: CUDA error: the provided PTX was compiled with an unsupported toolchain.
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
```

Appreciate you taking the time Craig, this is definitely an odd one.

**Comment by craigcitro** [comment]
"It doesn't work after a package install until I restart the runtime" is almost always an issue with the way python imports work: imports are cached (in `sys.modules`) and replacing contents on disk will often mean that some package `p` is loaded from one version, whereas `p.some.sub.package` is loaded from a different version, causing madness. Colab exacerbates this by only being able to install from a running python session.

OOC if you run just the pip install(s), do we warn you about needing to restart the runtime?

**Comment by craigcitro** [comment]
For anyone here for JAX issues: a quick workaround until we can get a fix out is to run

```
!pip install https://storage.googleapis.com/jax-releases/cuda111/jaxlib-0.1.71+cuda111-cp37-none-manylinux2010_x86_64.whl
```

and then restart your kernel (ctrl-m . or `Runtime` -> `Restart runtime`).

**Comment by Clevrai** [comment]
@craigcitro No warnings about restarting a runtime. Regardless of restarting or not, I still get the error, unfortunately.

**Comment by craigcitro** [comment]
@Clevrai I can repro with your example from [this comment](https://github.com/googlecolab/colabtools/issues/2452#issuecomment-978170707); my initial attempt to downgrade torch+friends fails with the same error, which makes me think the problem may be with one of the other packages in play here. You may need to do a little more digging (or asking upstream).

**Comment by melvin-dy-opt** [comment]
Hi, I've also been facing a similar error when loading pretrained models, both private pretrained models and publicly available models via HuggingFace. 

Minimal reproducible steps:
1. Prerequisite installs: 
```
!pip install --quiet ipadic
!pip install --quiet mecab-python3
!pip install --quiet fugashi
!pip install --quiet transformers

```
2. Acquire and load models into GPU
```
``` 
3.

**Comment by Clevrai** [comment]
Update on our end too: Craig's comment seems to work for all my examples too (as @melvin-dy-opt mentioned in the comment above). 

Thank you so much for your help @craigcitro! Hopefully, this can get patched up but this will do for now!

**Comment by x4nth055** [comment]
Yes, I had the same issue on A100, fixed it with @craigcitro 's workaround, thanks!

**Comment by YuyangLeng** [comment]
Had the same problem, thanks @craigcitro. 
!pip install https://storage.googleapis.com/jax-releases/cuda111/jaxlib-0.1.71+cuda111-cp37-none-manylinux2010_x86_64.whl
This fixed the issue.

**Comment by syskn** [comment]
We are getting the same error with A100.

Installing different torch/cuda versions did not work, but pinning jaxlib==0.1.71 and restarting the instance solved the issue as well - never thought this has to do with JAX. Thanks @craigcitro !

**Comment by jung-hunsoo** [comment]
I'm getting the same issue too. The error popped up when instantiating the model running `model_gan = MyGAN()`. No problem with CPU option. There's no `pip install _anything_` on the code.

Environment:
```
tf.__version__  # 2.7.0
```
MyGAN() inherits `tf.keras.Model` by calling `class MyGAN(tf.keras.Model):`.
![colab-a100-error](https://user-images.githubusercontent.com/6870447/143600579-e132b8c3-9ae8-4abe-9dc9-a4810f9dcfba.png)

I think downgrading CUDA may be helpful, but I have no idea which version to use and how to. Or, is there any way to choose another GPU accelerator, rather than A100?

**Comment by craigcitro** [comment]
This should be fixed -- if anyone's still hitting issues with A100, feel free to open a new issue.

**Comment by x4nth055** [comment]
Fun fact: Now I'm no longer getting an A100 after the fix.

**Comment by yaadhavraajagility** [comment]
How was this fixed? We are testing a custom version of libtorch compiled with `TORCH_CUDA_ARCH_LIST="6.1 7.5 8.7+PTX"` but it fails on the A100

**Comment by melvin-dy-opt** [comment]
@yaadhavraajagility  In case you skipped over it in the thread: https://github.com/googlecolab/colabtools/issues/2452#issuecomment-978204803

`!pip install https://storage.googleapis.com/jax-releases/cuda111/jaxlib-0.1.71+cuda111-cp37-none-manylinux2010_x86_64.whl`
-> Then restart your runtime.

Sometime in December 2021, the workaround in the comment above was no longer needed. If it's still occurring for you, even with the above workaround, it might be some other (similar) problem.

**Comment by WILLIAMGUTERRES** [comment]
Using GPU=T4, I added to ARCH, ARCH= -gencode arch=compute_75,code=[sm_75,compute_75]. So my Makefile I changed like that:
`ARCH= -gencode arch=compute_50,code=[sm_50,compute_50] \
             -gencode arch=compute_52,code=[sm_52,compute_52] \
	     -gencode arch=compute_61,code=[sm_61,compute_61] \
             -gencode arch=compute_75,code=[sm_75,compute_75]`
